smwe_rom/src/snes_utils/rom_slice.rs

```diff
@@ -0,0 +1,76 @@
+use std::{fmt, fmt::Formatter};
+
+use crate::snes_utils::addr::{Addr, AddrPc, AddrSnes};
+
+#[derive(Copy, Clone, Debug)]
+pub struct RomSlice<A: Addr> {
+    pub begin: A,
+    pub size:  usize,
+}
+
+pub type PcSlice = RomSlice<AddrPc>;
+pub type SnesSlice = RomSlice<AddrSnes>;
+
+impl PcSlice {
+    pub const fn new(begin: AddrPc, size: usize) -> Self {
+        Self { begin, size }
+    }
+}
+
+impl SnesSlice {
+    pub const fn new(begin: AddrSnes, size: usize) -> Self {
+        Self { begin, size }
+    }
+}
+
+impl<A: Addr> RomSlice<A> {
+    pub fn end(&self) -> A {
+        self.begin + self.size
+    }
+
+    pub fn shift_forward(self, offset: usize) -> Self {
```

Comment:
This naming is a bit confusing, shift makes me think of `<<`/`>>` operators rather than moving the beginning.
My suggestion for the API here would be:
```
pub fn offset(self, offset: isize) -> Self; // isize allows for negative offsets
pub fn skip(self, lengths: isize) -> Self; // unsure what the name should be here, but skip is not that bad
```

Reply:
I'm going to stick to `{offset,skip}_{forward,backward}` because using `isize` means adding more type conversions and adding another set of trait bounds to `Addr`.